### PR TITLE
Tensorflow Internal 86

### PR DESCRIPTION
**Pull request by stevenireeves**
This implements the rocm counter part to matrix_solve_op on GPU.  This PR will complete TF Internal 86. 

I interfaced with rocSolver to pull LU solver and back-substitution.

### CODE REVIEW

File: tensorflow/core/kernels/linalg/matrix_solve_op.cc [hunk 1]
```diff
@@ -114,9 +118,11 @@ class MatrixSolveOp : public LinearAlgebraOp<Scalar> {
   TF_DISALLOW_COPY_AND_ASSIGN(MatrixSolveOp);
 };
 
-#if GOOGLE_CUDA
+#if GOOGLE_CUDA || TESNORFLOW_USE_ROCM
```
Comment (deven-amd) [inline tensorflow/core/kernels/linalg/matrix_solve_op.cc]: typo --- `TESNORFLOW_USE_ROCM`  !!

that also implies the changes contained within are not getting picked up :(

File: tensorflow/core/kernels/linalg/matrix_solve_op.cc [hunk 2]
```diff
@@ -357,13 +364,261 @@ class MatrixSolveOpGpu : public AsyncOpKernel {
   bool adjoint_;
 };
 
+#elif TENSORFLOW_USE_ROCM
```
Comment (deven-amd) [inline tensorflow/core/kernels/linalg/matrix_solve_op.cc:367]: we need to merge the ROCm and CUDA implementations into one.

There are approx 250 lines of code here, out which maybe 15 are different between the ROCm and CUDA implementations.  We should be able to keep just those differences within `#if GOOGLE_CUDA .... #elif TENSORFLOW_USE_ROCM` blocks and keep the rest the same.

**Review by deven-amd — CHANGES_REQUESTED** [review]
General comments:
* Please rebase / cleanup the commit history to remove the merge commits.
* Enable the unit tests associated with this code, so that we know it is getting exercised, and that it is working correctly.

thanks

File: tensorflow/core/kernels/linalg/matrix_solve_op.cc (hunk 2 above)
Comment (stevenireeves) [reply tensorflow/core/kernels/linalg/matrix_solve_op.cc:367]: Thanks Deven, I'll get on that.

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**